src/lib/core/CHIPTLVReader.cpp

```diff
@@ -125,63 +125,94 @@ CHIP_ERROR TLVReader::Get(bool & v)
     return CHIP_NO_ERROR;
 }
 
+template<typename int_type>
+bool TLVReader::TLVTypeValid(uint64_t val64) const
+{
+    bool tlv_isSigned = GetType() == kTLVType_SignedInteger;
+
+    return tlv_isSigned == std::is_signed<int_type>::value &&
+           (tlv_isSigned ? CanCastTo<int_type>(CastToSigned(val64)) : CanCastTo<int_type>(val64));
+}
+
 CHIP_ERROR TLVReader::Get(int8_t & v)
 {
     uint64_t v64   = 0;
-    CHIP_ERROR err = Get(v64);
-    v              = CastToSigned(static_cast<uint8_t>(v64));
+
+    CHIP_ERROR err = _Get(v64);
+    ReturnErrorCodeIf(!TLVTypeValid<int8_t>(v64), CHIP_ERROR_WRONG_TLV_TYPE);
+    v = CastToSigned(static_cast<uint8_t>(v64));
```

Comment:
```suggestion
    v = static_cast<int8_t>(CastToSigned(v64));
```
is clearer.  But even clearer would be to have an overload of `_Get` that takes `int64_t` and checks for the SignedInteger() type before returning success.  Then here we could just `CanCastTo` and return with a cast...

Similar comments for all the signed types.

Reply:
This comment is still unaddressed.